### PR TITLE
Remove setting require_osd_release in post_upgrade step of ceph-mon

### PR DESCRIPTION
**Pull request by agileshaw**
- Require OSD release should happen as a post upgrade of ceph-osd and not ceph-mon. Therefore removed this step from the post-upgrade plan of CephMonApplication class (but keeping the one in pre-upgrade plan)
- Instead of comparing require_osd_release value with ceph-mon release, we compare it with the actual release of OSDs. This information is parsed from `ceph versions` command output running on a ceph-mon unit.

fixes: https://github.com/canonical/charmed-openstack-upgrader/issues/122

### CODE REVIEW

File: cou/utils/app_utils.py
Comment (gabrielcocenza) [inline cou/utils/app_utils.py:115]: Both `_get_required_osd_release` and ` _get_current_osd_release` they share the same structure:

1. run a command
2. check if the command returned 0
3. raise RunUpgradeError if not 0
4. raise  RunUpgradeError if JujuError happens.

Can't we share a single generic function that does this job?

File: cou/utils/app_utils.py
Comment (gabrielcocenza) [inline cou/utils/app_utils.py:168]: update docstrings

**Review by gabrielcocenza — CHANGES_REQUESTED** [review]

File: cou/utils/app_utils.py
Comment (agileshaw) [reply cou/utils/app_utils.py:128]: I actually didn't know it supports formatting to json. Thanks! Now parsing the output is much more robust.

File: cou/utils/app_utils.py
Comment (agileshaw) [reply cou/utils/app_utils.py:115]: They do have very similar structures, and I would love to merge them. However, their error messages are tailored, so it is hard to do so. Do you have any suggestions?

File: cou/utils/app_utils.py
Comment (agileshaw) [reply cou/utils/app_utils.py]: I thought about this, but doing regex matching seems to be a bit too much. In the latest patch, I changed it to use the location where the ceph release is, but I will think about the regex solution.

File: cou/utils/app_utils.py
Comment (gabrielcocenza) [reply cou/utils/app_utils.py:115]: taking a closer look, set_require_osd_release_option also has a similar structure. I would try something like:

```python
async def _run_ceph_mon_command(command, unit, model: COUModel, run_upgrade_err_msg):
    logger.debug("Running '%s' on '%s'", command, unit)
    try:
        result = await model.run_on_unit(unit_name=unit, command=command, timeout=600)
        if str(result["Code"]) != "0":
            raise RunUpgradeError(run_upgrade_err_msg) from CommandRunFailed(cmd=command, result=result)
        return result["Stdout"]

    except JujuError as exc:
        raise RunUpgradeError(run_upgrade_err_msg) from exc

async def _get_required_osd_release(unit: str, model: COUModel) -> str:
    cmd = "ceph versions -f json"
    err_msg = f"Cannot determine the current value of require_osd_release on ceph-mon unit '{unit}'."
    return json.loads(wait _run_ceph_mon_command(cmd, unit, err_msg)).get("require_osd_release", "")
```

This can increase substantially the readability and apply DRY

File: cou/utils/app_utils.py
Comment (gabrielcocenza) [inline cou/utils/app_utils.py:194]: I still think that a dictionary on openstack.py having the ceph  could be beneficial here.

```python
CEPH_RELEASES = OrderedDict(
    [
        ....
        ("pacific", VersionRange("16.2.0", "17.2.0" )),
        .....
    ]
)

....
if current_osd_release not in CEPH_RELEASES.keys():
     raise RunUpgradeError(f"Invalid ceph-osd release '{current_osd_release}'")
```

**Review by gabrielcocenza — CHANGES_REQUESTED** [review]

File: cou/utils/app_utils.py
Comment (agileshaw) [reply cou/utils/app_utils.py:95]: Updated

File: cou/utils/juju_utils.py
Comment (rgildein) [inline cou/utils/juju_utils.py:339]: Should we use same approach as we are doing in `run_action`? I mean to use `raise_on_failure`.

File: tests/unit/apps/test_auxiliary.py
Comment (rgildein) [inline tests/unit/apps/test_auxiliary.py:416]: not-blocking:
```suggestion
            description="Ensure require-osd-release option matches with ceph-osd version",
```

**Review by rgildein — COMMENTED** [review]
Mostly it looks good to me, but there is one comment about raising an exception.

File: cou/utils/juju_utils.py
Comment (agileshaw) [reply cou/utils/juju_utils.py:339]: I don't think so. The point of having `raise_on_failure` for actions is because we are retrying for getting action result. But for run on command, getting the result is suppose to be static and no need for retrying.

File: cou/utils/openstack.py
Comment (gabrielcocenza) [inline cou/utils/openstack.py:151]: I think it's possible to encounter clouds using `natilus` if they are coming from bionic, so let's add to this list

File: tests/unit/utils/test_app_utils.py
Comment (gabrielcocenza) [inline tests/unit/utils/test_app_utils.py:85]: add ("natilus", "octopus")

File: tests/unit/utils/test_app_utils.py
Comment (gabrielcocenza) [inline tests/unit/utils/test_app_utils.py:167]: If we are not parameterizing, why not use `octopus` directly here?

File: cou/utils/openstack.py
Comment (agileshaw) [reply cou/utils/openstack.py:151]: We are not letting people upgrade from bionic, so I don't think adding `nautilus` is necessary nor desirable

File: tests/unit/utils/test_app_utils.py
Comment (agileshaw) [reply tests/unit/utils/test_app_utils.py:167]: Because I think it it is clearer to see what we are expecting with `expected_osd_release`.

File: cou/utils/openstack.py
Comment (gabrielcocenza) [reply cou/utils/openstack.py:151]: I think it's easy to someone forget to do this post upgrade step of `require-osd-release` and have a focal cloud on natilus

Comment (agileshaw) [reply cou/utils/openstack.py:151]: We are only using this CEPH_RELEASES list for checking whether the osd versions are in these range,  and IMO if their cloud is correct, they should never have OSD running `nautilus` on focal. If the operator starts an upgrade with `nautilus` value of `required_osd_release` but osd is at `octopus`, that value will be overwrite to `octopus` in the pre-upgrade step of ceph-mon.

File: cou/utils/juju_utils.py
Comment (rgildein) [reply cou/utils/juju_utils.py:339]: But what if I want to run command and I expecting that it could failed?

Comment (agileshaw) [reply cou/utils/juju_utils.py:339]: Opened #148 to track

File: cou/utils/openstack.py
Comment (gabrielcocenza) [reply cou/utils/openstack.py:151]: I see your point and will resolve this conversation, but as you said `required_osd_release` will overwrite to octopus in the pre-upgrade step of ceph-mon, so I think it's good to have it on `test_set_require_osd_release_option_different_releases`

**Review by rgildein — APPROVED** [review]
LGTM

File: cou/utils/juju_utils.py
Comment (rgildein) [reply cou/utils/juju_utils.py:339]: Thanks

**Review by gabrielcocenza — APPROVED** [review]
LGTM. Thanks

**Review by rgildein — APPROVED** [review]
LGTM

**Comment by agileshaw** [comment]
Will wait for the CI to trigger again before merging

**Pull request merged**
